### ISSUE TITLE
Double scaling issue in 04_ATT_LSTM_model_template

### ISSUE THREAD

**Issue by miohtama** [FranQuant/AI-based-Trading-Strategies#9]
Hi,

After scratching my head for many days, I think I have identified an issue that affects `04_ATT_LSTM_model_template` and maybe some other notebooks: The input data is scaled twice.

## Scaling 1: The first scaler is created for overall training and test split.

Set by `train_cutoff = '2022-01-01'`: this is overall scaling for the whole pricing history.

```python
#  Step 1: Define a chronological cutoff date for train/test split
train_cutoff = '2022-01-01'  # Modify if needed

df_train = df.loc[:train_cutoff]
df_test = df.loc[train_cutoff:]

X_train = X.loc[df_train.index]
y_train = y.loc[df_train.index]
X_test = X.loc[df_test.index]
y_test = y.loc[df_test.index]
```

And then this is scaled to:

```python
scaler = StandardScaler()
scaler.fit(X_train_selected)  # Fit only on training data

X_train_scaled = scaler.transform(X_train_selected)
X_test_scaled = scaler.transform(X_test_selected)  # Use same scaler on test set

# Convert back to DataFrame for consistency
X_train_scaled = pd.DataFrame(X_train_scaled, index=X_train_selected.index, columns=X_train_selected.columns)
X_test_scaled = pd.DataFrame(X_test_scaled, index=X_test_selected.index, columns=X_test_selected.columns)
```

## Scaling 2: When doing LSTM sequencing, it will read `X_test_scaled` 

```python
# Generate training & test sequences
X_seq_train, y_seq_train = create_sequences(X_train_scaled, y_train, seq_length=sequence_length)
X_seq_test, y_seq_test = create_sequences(X_test_scaled, y_test, seq_length=sequence_length)
```

BUT 

There is another scaler in the walk-forward folding loop, which reads the sequenced data (already scaled) and scales the already scaled-sequenced again:

```python
for fold, (train_index, test_index) in enumerate(tscv.split(X_seq_train)):
    print(f"\nProcessing Fold {fold + 1}/{tscv.n_splits}...")

    # Train/test split for current fold
    X_train_fold = X_seq_train[train_index]
    X_test_fold = X_seq_train[test_index]
    y_train_fold = y_seq_train[train_index]
    y_test_fold = y_seq_train[test_index]

    # Scaling (fold-by-fold)
    nsamples_train, timesteps, n_features = X_train_fold.shape
    X_train_fold_2d = X_train_fold.reshape(-1, n_features)

    scaler = StandardScaler()
    X_train_fold_scaled = scaler.fit_transform(X_train_fold_2d).reshape(X_train_fold.shape)
    X_test_fold_scaled = scaler.transform(X_test_fold.reshape(-1, n_features)).reshape(X_test_fold.shape)
```

This does not seem to affect the model performance (you still get accuracy). But I doubt this is intentional? Probably scaling already scaled data does little, but because the period is different, it may twist some things around.

My questions are 
- Which scaling should be preferred? My own guess is that per-fold scaling should give better results because not all indicators are stationary and prices shift over time.
- If this is unintentional, would you like me to submit a PR fix?

**Comment by FranQuant** [comment]
Hi Mikko,

Thanks a lot for taking the time to review the repo and point out the
potential double-scaling in the 04_ATT_LSTM_model_template notebook. I
really appreciate the detailed breakdown you shared in the issue.

I’m currently reviewing the repository with some additional tooling to make
sure the pipeline is clean and consistent across notebooks. Before making
changes to the main branch, I’d like to carefully test the implications of
fold-by-fold scaling vs. global scaling, especially in relation to
stationarity and potential lookahead effects.

Also, you mentioned that you got the example attention notebook working
with BTC data. Would you mind sharing the dataset or the way you set it up?
That would be very helpful to cross-check results and ensure we’re looking
at things under the same conditions.

Thanks again for your thorough review and for stress-testing the notebooks
— really valuable.

Best,
Francisco

**Comment by FranQuant** [comment]
Thanks for catching this — you were right that the notebook was unintentionally **scaling twice**. I’m standardizing on this policy (to avoid leakage and to respect non-stationarity):

1. **No scaling before sequencing.** Build sequences from raw features.  
2. **Per-fold scaling in CV.** Inside each time-series fold: fit the scaler on the fold’s *train* sequences (2D reshape), then transform both fold-train and fold-val.  
3. **Single final OOS fit.** After CV, fit the scaler once on all training sequences, transform the post-cutoff test sequences once, and evaluate OOS.  
4. **Model saving.** Save the OOS model (or best-fold weights), not just “last fold.”  

This way the mean/variance estimates are aligned with the actual information set available at each point in time — important because most indicators aren’t strictly stationary.

**Notebook changes (summary):**
- Remove the early `StandardScaler` on `X_train_selected/X_test_selected`.  
- Keep per-fold scaling only (fit on fold-train → transform fold-val).  
- Add a final OOS step that fits on all training sequences and predicts on the test set once.  
- Save the OOS model (or best-fold weights) rather than the last CV fold.  

Minimal code sketch of the intended flow:

```python
# --- Create sequences from UNscaled features ---
X_seq_train, y_seq_train = create_sequences(X_train_selected, y_train, seq_len)
X_seq_test,  y_seq_test  = create_sequences(X_test_selected,  y_test,  seq_len)

# --- Inside each CV fold ---
ns, T, F = X_train_fold.shape
scaler = StandardScaler()
X_train_fold_scaled = scaler.fit_transform(X_train_fold.reshape(-1, F)).reshape(ns, T, F)
X_val_fold_scaled   = scaler.transform(X_val_fold.reshape(-1, F)).reshape(X_val_fold.shape)

# --- Final OOS step ---
scaler_final = StandardScaler()
Xtr2d = X_seq_train.reshape(-1, F)
Xte2d = X_seq_test.reshape(-1, F)
Xtr_scaled = scaler_final.fit_transform(Xtr2d).reshape(X_seq_train.shape)
Xte_scaled = scaler_final.transform(Xte2d).reshape(X_seq_test.shape)

Appreciate you catching this — it’s a helpful review that makes the pipeline more robust.

**Comment by miohtama** [comment]
🤝 Likewise. Appreciate you for all hard work you have open sourced.
